### ISSUE TITLE
Provide OpenModelica system libraries for MSL 3.2.3 and 4.0.0

### ISSUE THREAD

**Issue by christiankral** [OpenModelica/OpenModelica#7376]
I am currently providing some open source libraries which are included as system libraries in OpenModelica. Currently all these libraries are based on the MSL 3.2.3 and stored on a GitHub branch named `release`. 

I wonder if there is a way to provide the same library both for the MSL 3.2.3 and 4.0.0. If yes, how are the library files provided (in which branches)?

**Comment by adrpo** [comment]
Should be OK if you have the proper uses annotations to MSL and the libraries have different version numbers.
We start using the Package Manager for libraries but we don't have a GUI yet for it. You can find the index of libraries here:
https://github.com/OpenModelica/OMPackageManager/blob/master/repos.json
@sjoelund how do you add a new library to the package manager index? Open a PR?

**Comment by sjoelund** [comment]
Yes, opening a PR. I guess the libraries are already there and there is just the matter of adding the libraries.

If they have the same version number, loadModel will not really know how to load something that has a uses annotation on this library though. But this is something that can be resolved. Of course, we support conversion scripts too with API. So hopefully we can just silently upgrade libraries soon :)

**Comment by christiankral** [comment]
I am afraid my problem is to properly understand the syntax of [repos.json](https://github.com/OpenModelica/OMPackageManager/blob/master/repos.json). 

```
  "ElectroMechanicalDrives": {
    "names": ["ElectroMechanicalDrives"],
    "github": "christiankral/ElectroMechanicalDrives",
    "support": [
      ["prerelease", "noSupport"],
      [">=2.2.0", "experimental"],
      ["*", "obsolete"]
    ]
```

1. Does this mean that any version >=2.2.0 is provided through the OM package manager?
2. How are versions =>2.2.0 obtained from Github by the package manager? Example (these tags and branches do not yet exist on the repo, however): I want to provide the versions 2.2.0, 2.2.1 and 2.3.0
    a. OMPackageManager searches and finds the tags, e.g., v2.2.0, v2.2.1, v2.3.0, ...? 
    b. OMPackageManager searches and finds the branches, e.g., 2.2.0, 2.2.1, 2.3.0, ...?
3. What does `"experimental"` stand for?
4. what does `["*", "obsolete"]` stand for?

Is there a specification of the json file?

**Comment by sjoelund** [comment]
1. All versions are provided, but the best supported version is installed by default.
2. It searches for all tags (and branches explicitly given as "branches": [...] (see other entries)
3. experimental is sorted lower than support, but higher than noSupport and obsolete. Things marked fullSupport have a commitment to work in OM. support means we should answer tickets because most things are expected to work (or that the library developer tests it in OM).
4. It's a wildcard. Any version not marked as prerelease (like a tag v2.3.0-beta.2, or a branch v2.3.0-master), and not `>=2.2.0` will match anything there (so old releases will be marked obsolete)

I thought there was a specification, but I guess not :)

**Comment by christiankral** [comment]
In order to create a proper PR for the entry of the ElectroMechanicalDrives library I had a look at the Buldings library in [repos.json](): 

```json
"Buildings": {
  "names": ["Buildings"],
  "github": "lbl-srg/modelica-buildings",
  "branches": {"master": "master"},
  "support": [
    ["prerelease", "noSupport"],
    [">=5.1.0", "support"],
    ["*", "obsolete"]
  ]
},
```

1. I checked the tagged versions on the [Buldings repsotory](https://github.com/lbl-srg/modelica-buildings/tags):
![image](https://user-images.githubusercontent.com/4184218/126786446-f7e5cabf-8be7-442b-b42d-aaf435c34aab.png)
2. From the entry in the json file and the proved tags I would expect to find the versions marked with red color in the nightly OM Ubuntu 20.04 repository:
![image](https://user-images.githubusercontent.com/4184218/126786980-192d7b6c-0d47-4dac-8ecd-612b219c47bf.png)

- Only v6.0.0 and v7.0.0 are available 
- even though `>=5.1.0` is specified in the json file. 

Why?

**Comment by christiankral** [comment]
The same issue applies (on my understanding) for the [ElectroMechanicalDrives](https://github.com/christiankral/ElectroMechanicalDrives/tags) library which has the following version tags:

![image](https://user-images.githubusercontent.com/4184218/127180200-196d2630-f59c-4089-b749-caaf9bc879b5.png)

Yet, in the OM repository only v2.2.0 is provided:

![image](https://user-images.githubusercontent.com/4184218/127180271-8ff923be-5e3f-4baf-a086-3d5dcadc49bf.png)

Comments and support is appreciated.

**Comment by casella** [comment]
@sjoelund is currently on vacation, I hope he can sort this out when he comes back. A readme.md file with a few lines of spec would indeed help 😄

**Comment by sjoelund** [comment]
It won't become visible in the Debian/Ubuntu package manager. It becomes visible in the OpenModelica package manager.
For the next release, all of the libraries are expected to disappear.

**Comment by christiankral** [comment]
To understand that correctly: For the next release of OpenModelica each library version listed in [repos.json](https://github.com/OpenModelica/OpenModelica/issues/7376) 

1. is distributed with the released OpenModelica binary?
2. can be loaded (manually) through the PackageManager of OpenModelica?

Which one applyies, 1. or 2.?

**Comment by sjoelund** [comment]
2. No library will be distributed (or very few). This means you will need an internet connection to download new libraries

**Comment by christiankral** [comment]
I managed to use the OM PackageManager in OM 1.18.0 to install different versions of some system libraries. I understand, that one day a GUI will be available to manage such Modelica packages. 

Particularly now, as the MSL 3.2.3 and 4.0.0 are often used one after the other in different projects, I wonder if it were also possible to include all the installed versions in the "Systems Library" Menu somehow. I am thinking of something like an additional submenu associated with each "System Library" that the shows the installed versions to pick from. 

If I am working on an MSL 3.2.3 project, loading, for example, the "ElectroMechanicalDrives" library, the version 3.0.0 is loaded, even though "ElectroMechanicalDrives 3.0.0" has an explicit annotation indicating that it requires the MSL 4.0.0. Yet I have no option to explicitly load "ElectroMechanicalDrives 2.4.0" even though it is installed. 

![image](https://user-images.githubusercontent.com/4184218/134564322-d0f801a8-adbd-43ed-b5b2-a61f158d5413.png)

I can only specify the "System Library" version number in the Options menu:

![image](https://user-images.githubusercontent.com/4184218/134563556-78f283d4-7fbe-44ff-95e0-85fff9f036c8.png)

**Comment by casella** [comment]
@christiankral, yes, we're trying to have a GUI for the next version, though it turns out getting a good specification for it is far from trivial.

See also the discussion in [MSL #3003](https://github.com/modelica/ModelicaSpecification/issues/3003)

@sjoelund, can we start improving the loader logic of OMC so that it looks ahead for dependencies and tries to avoid this kind of situations?

**Comment by sjoelund** [comment]
> @sjoelund, can we start improving the loader logic of OMC so that it looks ahead for dependencies and tries to avoid this kind of situations?

We could if someone has the time for it. The same logic could be used to improve the installation of packages. It is using a greedy algorithm right now, choosing the latest packages of libraries one by one.

**Comment by casella** [comment]
See also #8060. @christiankral we are trying to get this into 1.19.0

**Comment by christiankral** [comment]
The actual nightly implementation already looks quite good!

**Comment by casella** [comment]
@christiankral, good to hear! Do you have any further suggestions or requirements, or can we close this ticket?

**Comment by christiankral** [comment]
If a few comments:

1. I have one comment on installing large libraries:

![image](https://user-images.githubusercontent.com/4184218/142741562-067fa3f3-9cdf-4693-8c51-34fc4377866b.png)

After pressing *OK* the installation process starts but the user has no feedback about the status of the installation process. After installing the library the Messages Browser reports that the library has been installed, but before this message appears, the user has no feedback at all.

2. Is there way to remove (uninstall) a particular version of a library?

3. If I click on *Upgrade Installed Libraries* a error message is caused:

```
[10] 22:41:25 Scripting Error
[/home/christian/.openmodelica/libraries/index.json: 0:0-0:0]: The package index did not contain an entry for package ApproxSpline of version 1.0.1. There are other versions that claim to be compatible: 1.0.1-master.
```

This is a list of the libraries installed on my computer:

![image](https://user-images.githubusercontent.com/4184218/142741720-8da71f3e-3bec-4cbc-a8dd-5dc77b6e2660.png)

**Comment by casella** [comment]
@christiankral, this last issue is covered by #8187. I think we can close this one, as what you were asking for originally is now provided in the nightly build, and will be included in the 1.19.0 release.

**Comment by christiankral** [comment]
@casella Yes this is all good. Thank you.
